src/transformers/models/ngen3/modular_ngen3.py

```diff
@@ -0,0 +1,168 @@
+
+def apply_rotary_pos_emb(q: torch.Tensor, k: torch.Tensor) -> tuple:
+    """
+    Applies rotary positional embeddings (RoPE) to query and key tensors.
+    
+    Args:
+        q (torch.Tensor): Tensor of shape (B, n_head, T, head_dim)
+        k (torch.Tensor): Tensor of shape (B, n_head, T, head_dim)
+    
+    Returns:
+        Tuple of rotated (q, k)
+    """
+    T = q.size(-2)
+    dim = q.size(-1)
+    device = q.device
+    inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2, device=device).float() / dim))
+    positions = torch.arange(T, device=device).float()
+    sinusoid_inp = torch.einsum("i,j->ij", positions, inv_freq)
+    sin = torch.sin(sinusoid_inp).unsqueeze(0).unsqueeze(0)
+    cos = torch.cos(sinusoid_inp).unsqueeze(0).unsqueeze(0)
+    q1, q2 = q[..., :dim//2], q[..., dim//2:]
+    k1, k2 = k[..., :dim//2], k[..., dim//2:]
+    q_rot = torch.cat((q1 * cos - q2 * sin, q2 * cos + q1 * sin), dim=-1)
+    k_rot = torch.cat((k1 * cos - k2 * sin, k2 * cos + k1 * sin), dim=-1)
+    return q_rot, k_rot
+
+class NGEN3(nn.Module):
+    def __init__(self, config: NGEN3Config):
+        super().__init__()
+        self.config = config
+        self.tok_emb = nn.Embedding(config.vocab_size, config.n_embd)
+        self.drop = nn.Dropout(config.dropout)
+        self.blocks = nn.Sequential(*[Block(config) for _ in range(config.n_layer)])
+        self.ln_f = nn.LayerNorm(config.n_embd)
+        if config.instruct:
+            self.instruct_dense = nn.Linear(config.n_embd, config.n_embd)
+        self.head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
+        self.apply(self._init_weights)
+    
+    def _init_weights(self, module):
+        if isinstance(module, (nn.Linear, nn.Embedding)):
+            nn.init.normal_(module.weight, mean=0.0, std=0.02)
+        if isinstance(module, nn.Linear) and module.bias is not None:
+            nn.init.zeros_(module.bias)
+    
+    def forward(self, input_ids: torch.Tensor, instruct_mode: bool = False) -> torch.Tensor:
+        x = self.tok_emb(input_ids)
+        x = self.drop(x)
+        x = self.blocks(x)
+        x = self.ln_f(x)
+        if self.config.instruct and instruct_mode:
+            x = self.instruct_dense(x)
+        logits = self.head(x)
+        return logits
+
+class Block(nn.Module):
+    def __init__(self, config: NGEN3Config):
+        super().__init__()
+        self.ln1 = nn.LayerNorm(config.n_embd)
+        self.attn = CausalSelfAttention(config)
+        self.ln2 = nn.LayerNorm(config.n_embd)
+        self.mlp = MoEMLP(config) if config.use_moe else MLP(config)
+    
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Residual connection for attention
+        residual = x
+        x = self.ln1(x)
+        x = self.attn(x)
+        x = residual + x
+        # Residual connection for feedforward
+        residual = x
+        x = self.ln2(x)
+        x = self.mlp(x)
+        return residual + x
+
+class CausalSelfAttention(nn.Module):
```

Comment:
this should be completely equivalent to a variation of LlamaAttention with fused qkv. Nothing very new and the mask should not be saved there